libnmstate/nm/plugin.py

```diff
@@ -0,0 +1,204 @@
+            iface_info[Interface.IPV4] = nm_ipv4.get_info(act_con)
+            iface_info[Interface.IPV6] = nm_ipv6.get_info(act_con)
+            iface_info.update(nm_wired.get_info(dev))
+            iface_info.update(nm_user.get_info(self.client, dev))
+            iface_info.update(nm_vlan.get_info(dev))
+            iface_info.update(nm_vxlan.get_info(dev))
+            iface_info.update(nm_bridge.get_info(self.client, dev))
+            iface_info.update(nm_team.get_info(dev))
+
+            if nm_bond.is_bond_type_id(type_id):
+                bondinfo = nm_bond.get_bond_info(dev)
+                iface_info.update(_ifaceinfo_bond(bondinfo))
+            elif nm_ovs.CAPABILITY in self.capabilities:
+                if nm_ovs.is_ovs_bridge_type_id(type_id):
+                    iface_info["bridge"] = nm_ovs.get_ovs_info(
+                        self.client, dev, devices_info
+                    )
+                    iface_info = _remove_ovs_bridge_unsupported_entries(
+                        iface_info
+                    )
+                elif nm_ovs.is_ovs_port_type_id(type_id):
+                    continue
+
+            info.append(iface_info)
+
+        info.sort(key=itemgetter("name"))
+
+        return info
+
+    def get_routes(self):
+        return {
+            Route.RUNNING: (
+                nm_ipv4.get_route_running(self.client)
+                + nm_ipv6.get_route_running(self.client)
+            ),
+            Route.CONFIG: (
+                nm_ipv4.get_route_config(self.client)
+                + nm_ipv6.get_route_config(self.client)
+            ),
+        }
+
+    def get_route_rules(self):
+        return {
+            RouteRule.CONFIG: (
+                nm_ipv4.get_routing_rule_config(self.client)
+                + nm_ipv6.get_routing_rule_config(self.client)
+            )
+        }
+
+    def get_dns_client_config(self):
+        return {
+            DNS.RUNNING: nm_dns.get_running(self.client),
+            DNS.CONFIG: nm_dns.get_config(
+                nm_ipv4.acs_and_ip_profiles(self.client),
+                nm_ipv6.acs_and_ip_profiles(self.client),
+            ),
+        }
+
+    def get_checkpoint(self):
+        return str(self._checkpoint.dbuspath) if self._checkpoint else None
+
+    def _load_checkpoint(self, checkpoint_path):
+        all_checkpoint_paths = get_checkpoints(self.client)
+        if not all_checkpoint_paths or (
+            checkpoint_path and checkpoint_path not in all_checkpoint_paths
+        ):
+            raise NmstateValueError(
+                f"Specified checkpoint {checkpoint_path} does not exists"
+            )
+        self._checkpoint = CheckPoint(
+            self._ctx,
+            dbuspath=checkpoint_path
+            if checkpoint_path
+            else all_checkpoint_paths[0],
+        )
+
+    def create_checkpoint(self, rollback_timeout):
+        if self._checkpoint:
+            raise NmstateConflictError("Only single checkpoint is allowed")
```

Comment:
This is new AFAICS, before an user could apply a state and get the checkpoint, then apply a new state and rollback to the latest one if I am not wrong. Now, will nmstate raises an exception if the user does not commit the current state before setting the new one? Sorry if I am missing something.

Reply:
Maybe an option could be to keep a list of checkpoints. If the user specifies a checkpoint to rollback, everything is fine if not, then rollback to the last one. And destroy all the other checkpoints.

Reply:
Supporting a list checkpoints could be an RFE.

The reason we enforce single checkpoint is to make sure only single nmstate is editing the system, otherwise unexpected behavior might be happen(some remove slave, some add slave).

And clearly this is unrelated to this PR.

Reply:
The flags we used in the past already ensured that there will be only one checkpoint at a time, so it is nothing new to support only one checkpoint.

Reply:
But what I meant is: before nmstate was not raising an exception when using 2 checkpoints. Now yes. I am fine with that as we are not supporting it. But then I think we should document it somewhere.

Reply:
The document is https://www.nmstate.io/devel/api#nmstateconflicterror

Reply:
Oh. Thanks for explaining, seems fine to me now.